_posts/2021-10-25-clean-architecture-story.md

```diff
@@ -0,0 +1,332 @@
+— with definition of schedule creator if required — without any irrelevant details, like database, UI, framework and so
+on. Test only business rules, without unnecessary details.
+
+## Reserving the slot
+To finish the reservation we have to add at least one more use case — one for reservation of a free slot. Provided that
+we re-use existing logic, the interaction is still simple:
+
+```kotlin
+fun reserve(slotId: SlotId): DaySchedule {
+  val daySchedule = getScheduleUseCase.getSchedule(scheduleDay = slotId.day)
+
+  val modifiedSchedule = daySchedule.reserveSlot(slotId.index)
+
+  return dayScheduleRepository.save(modifiedSchedule)
+}
+```
+
+(full commit: [GitHub](https://github.com/michal-kowalcze/clean-architecture-example/commit/7b7961b28107c3c89d40ce69a8383bf9f32337b0))
+
+And, as we can see — the slot reservation business rule (and constraint) is implemented at the domain model itself — so
+we are safe, that any other interaction, any other use case, is not going to break these rules. This approach also
+simplifies testing, as business rules can be verified in separation from the use case interaction logic.
+
+## Where is the “Clean Architecture”?
+Let‘s stop with business logic for a moment. We created quite thoughtful, extensible code for sure, but why are we
+talking about “Clean” architecture? We already used Domain-Driven Design and Hexagonal architecture concepts. Is there
+anything more? Imagine that another person is going to help us with implementation. She is not aware of the source code
+yet and simply would like to take a look at the codebase. And she sees: ![Use case classes](/img/articles/2021-10-25-clean-architecture-story/use_case_classes.png)
+It looks like something to her, doesn‘t it? A kind of reservation system! It is not yet another domain service with
+some methods that have no clear connection with possible uses — the list of classes itself describes what the system
+can do.
+
+## The first assumption
+We have a mocked implementation as the schedule creator. It is OK to test logic at the unit test level, but not enough
+to run a prototype.
+
+After a short call with our customer we know more about the daily schedule — there are six slots, two hours each,
+starting at 8:oo a.m. We also know that this recipe for the daily schedule is very, very simple and it is going to be
+changed soon (e.g. to accommodate for holidays, etc.). All these issues will be solved later, now we are in the
+prototype stage and our desired outcome is to have a working demo for our stranger.
+
+Where to put this simple implementation of the schedule creator?So far, the domain used an interface for that. Are we
+going to put an implementation of this interface to the infrastructure package and treat it as something outside the
+domain? Certainly not! It is simple but this is part of the domain itself, we simply replace the interface with class
```

Comment:
Unikałbym słowa `simple` bo dla kogoś to o czym piszesz wcale łatwe być nie musi, mamy też jego powtórzenie. Prędzej poszedłbym w stronę np. `not complicated`.

Musiałem też przeczytać kilka razy to zdanie, żeby zrozumieć że odnosi się do interfejsu schedule creatora.

Reply:
Lekko zmieniłem, powinno być teraz ciut łatwiej złapać wątek, bo nawiązanie jest wprost.